pkg/openstack/actuator.go

```diff
@@ -0,0 +1,260 @@
+	sLog := logger.WithFields(log.Fields{
+		"targetSecret": fmt.Sprintf("%s/%s", cr.Spec.SecretRef.Namespace, cr.Spec.SecretRef.Name),
+		"cr":           fmt.Sprintf("%s/%s", cr.Namespace, cr.Name),
+	})
+
+	if existingSecret == nil || existingSecret.Name == "" {
+		if clouds == "" {
+			msg := "new access key secret needed but no key data provided"
+			sLog.Error(msg)
+			return &actuatoriface.ActuatorError{
+				ErrReason: minterv1.CredentialsProvisionFailure,
+				Message:   msg,
+			}
+		}
+		sLog.Info("creating secret")
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cr.Spec.SecretRef.Name,
+				Namespace: cr.Spec.SecretRef.Namespace,
+				Annotations: map[string]string{
+					minterv1.AnnotationCredentialsRequest: fmt.Sprintf("%s/%s", cr.Namespace, cr.Name),
+				},
+			},
+			Data: map[string][]byte{"clouds.yaml": []byte(clouds)},
+		}
+
+		err := a.Client.Create(context.TODO(), secret)
+		if err != nil {
+			sLog.WithError(err).Error("error creating secret")
+			return err
+		}
+		sLog.Info("secret created successfully")
+		return nil
+	}
+
+	// Update the existing secret:
+	sLog.Debug("updating secret")
+	origSecret := existingSecret.DeepCopy()
+	if existingSecret.Annotations == nil {
+		existingSecret.Annotations = map[string]string{}
+	}
+	existingSecret.Annotations[minterv1.AnnotationCredentialsRequest] = fmt.Sprintf("%s/%s", cr.Namespace, cr.Name)
+	if clouds != "" {
+		existingSecret.Data["clouds.yaml"] = []byte(clouds)
+	}
+
+	if !reflect.DeepEqual(existingSecret, origSecret) {
+		sLog.Info("target secret has changed, updating")
+		err := a.Client.Update(context.TODO(), existingSecret)
+		if err != nil {
+			msg := "error updating secret"
+			sLog.WithError(err).Error(msg)
+			return &actuatoriface.ActuatorError{
+				ErrReason: minterv1.CredentialsProvisionFailure,
+				Message:   msg,
+			}
+		}
+	} else {
+		sLog.Debug("target secret unchanged")
+	}
+
+	return nil
+}
+
+func (a *OpenStackActuator) loadExistingSecret(cr *minterv1.CredentialsRequest) (*corev1.Secret, string) {
+	logger := a.getLogger(cr)
+	var existingClouds string
+
+	// Check if the credentials secret exists, if not we need to inform the syncer to generate a new one:
+	existingSecret := &corev1.Secret{}
+	err := a.Client.Get(context.TODO(), types.NamespacedName{Namespace: cr.Spec.SecretRef.Namespace, Name: cr.Spec.SecretRef.Name}, existingSecret)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Debug("secret does not exist")
+		}
+	} else {
+		keyBytes, ok := existingSecret.Data["clouds.yaml"]
+		if !ok {
+			logger.Warning("secret did not have expected key: clouds.yaml, will be regenerated")
+		} else {
+			decoded := string(keyBytes)
+			existingClouds = string(decoded)
+			logger.Debug("found clouds.yaml in target secret")
+		}
+	}
+	return existingSecret, existingClouds
+}
+
+func (a *OpenStackActuator) getCloudCredentialsSecret(ctx context.Context, logger log.FieldLogger) (*corev1.Secret, string, error) {
```

Comment:
Could you rename this to getRootCloudCredentialsSecret.

Reply:
Done